src/nspawn/nspawn.c

```diff
@@ -4089,7 +4153,24 @@ static int outer_child(
         if (r < 0)
                 return r;
 
-        r = copy_devnodes(directory);
+        /* FUSE is only userns-safe in FUSE version 7.27 and later.
+         * https://github.com/torvalds/linux/commit/da315f6e03988a7127680bbc26e1028991b899b8 */
+        uint32_t fuse_major, fuse_minor;
+        r = get_fuse_version(&fuse_major, &fuse_minor);
+        if (r < 0)
+                log_warning_errno(r, "Disabling FUSE: Failed to determine FUSE version: %m");
+        else if (r == 0)
+                log_debug("Disabling FUSE: FUSE appears to be disabled on the host");
+        else if (fuse_major < 7 || (fuse_major == 7 && fuse_minor < 27))
+                log_debug("Disabling FUSE: FUSE version %" PRIu32 ".%" PRIu32 " is too old to support user namespaces",
+                          fuse_major, fuse_minor);
+        else
+                enable_fuse = true;
+        l = send(fd_outer_socket, &enable_fuse, sizeof enable_fuse, 0);
+        if (l < 0)
+                return log_error_errno(errno, "Failed to send whether to enable FUSE: %m");
```

Comment:
The function outer_child() is already too long. I'd like to split out this as `send_fuse_supported()` or so, which return a tristate.

Reply:
I can see the case for a `bool should_enable_fuse(void)`, but send/recv between outer_child() and inner_child() is a core-enough part of their dance that pulling the send() itself out feels wrong.

Reply:
With this and the below comment about `ERRNO_IS_DEVICE_ABSENT`, I could do this; but is it actually better, or just different?

```patch
--- a/src/nspawn/nspawn.c
+++ b/src/nspawn/nspawn.c
@@ -2147,11 +2147,6 @@ static int setup_boot_id(void) {
         return mount_nofollow_verbose(LOG_ERR, NULL, to, NULL, MS_BIND|MS_REMOUNT|MS_RDONLY|MS_NOSUID|MS_NOEXEC|MS_NODEV, NULL);
 }
 
-/* Returns:
- *  - <0 (-errno) : major/minor are unmodified : failed to determine the FUSE version
- *  - 0           : major/minor = 0/0          : FUSE does not appear to be present
- *  - 1           : major/minor are set        : successfully determined the FUSE version
- */
 static int get_fuse_version(uint32_t *ret_major, uint32_t *ret_minor) {
         /* Must be called with mount privileges, either via arg_privileged or by being uid=0 in new
          * CLONE_NEWUSER/CLONE_NEWNS namespaces. This is true when called from outer_child(). */
@@ -2178,10 +2173,6 @@ static int get_fuse_version(uint32_t *ret_major, uint32_t *ret_minor) {
         /* Get a FUSE handle. */
         fuse_fd = open("/dev/fuse", O_CLOEXEC|O_RDWR);
         if (fuse_fd < 0) {
-                if (ERRNO_IS_DEVICE_ABSENT(errno)) {
-                        *ret_major = *ret_minor = 0;
-                        return 0;
-                }
                 return log_debug_errno(errno, "Failed to open /dev/fuse: %m");
         }
         if (asprintf(&opts, "fd=%i,rootmode=40000,user_id=0,group_id=0", fuse_fd) < 0)
@@ -2206,7 +2197,31 @@ static int get_fuse_version(uint32_t *ret_major, uint32_t *ret_minor) {
                 return log_debug_errno(SYNTHETIC_ERRNO(EIO), "Initial FUSE_INIT request from /dev/fuse is too short");
         *ret_major = request.structured.body.major;
         *ret_minor = request.structured.body.minor;
-        return 1;
+        return 0;
+}
+
+static bool should_enable_fuse(void) {
+        uint32_t fuse_major, fuse_minor;
+        int r;
+
+        r = get_fuse_version(&fuse_major, &fuse_minor);
+        if (r < 0) {
+                if (ERRNO_IS_NEG_DEVICE_ABSENT(r))
+                        log_debug_errno(r, "Disabling FUSE: FUSE appears to be disabled on the host: %m");
+                else
+                        log_warning_errno(r, "Disabling FUSE: Failed to determine FUSE version: %m");
+                return false;
+        }
+
+        /* FUSE is only userns-safe in FUSE version 7.27 and later.
+         * https://github.com/torvalds/linux/commit/da315f6e03988a7127680bbc26e1028991b899b8 */
+        if (fuse_major < 7 || (fuse_major == 7 && fuse_minor < 27)) {
+                log_debug("Disabling FUSE: FUSE version %" PRIu32 ".%" PRIu32 " is too old to support user namespaces",
+                          fuse_major, fuse_minor);
+                return false;
+        }
+
+        return true;
 }
 
 static int copy_devnodes(const char *dest, bool enable_fuse) {
@@ -3870,7 +3885,7 @@ static int outer_child(
         _cleanup_(bind_user_context_freep) BindUserContext *bind_user_context = NULL;
         _cleanup_strv_free_ char **os_release_pairs = NULL;
         _cleanup_close_ int fd = -EBADF, mntns_fd = -EBADF;
-        bool idmap = false, enable_fuse = false;
+        bool idmap = false, enable_fuse;
         const char *p;
         pid_t pid;
         ssize_t l;
@@ -4153,19 +4168,7 @@ static int outer_child(
         if (r < 0)
                 return r;
 
-        /* FUSE is only userns-safe in FUSE version 7.27 and later.
-         * https://github.com/torvalds/linux/commit/da315f6e03988a7127680bbc26e1028991b899b8 */
-        uint32_t fuse_major, fuse_minor;
-        r = get_fuse_version(&fuse_major, &fuse_minor);
-        if (r < 0)
-                log_warning_errno(r, "Disabling FUSE: Failed to determine FUSE version: %m");
-        else if (r == 0)
-                log_debug("Disabling FUSE: FUSE appears to be disabled on the host");
-        else if (fuse_major < 7 || (fuse_major == 7 && fuse_minor < 27))
-                log_debug("Disabling FUSE: FUSE version %" PRIu32 ".%" PRIu32 " is too old to support user namespaces",
-                          fuse_major, fuse_minor);
-        else
-                enable_fuse = true;
+        enable_fuse = should_enable_fuse();
         l = send(fd_outer_socket, &enable_fuse, sizeof enable_fuse, 0);
         if (l < 0)
                 return log_error_errno(errno, "Failed to send whether to enable FUSE: %m");
```

Reply:
Backed out.